apecs/src/Apecs/Core.hs

```diff
@@ -62,13 +64,16 @@ class Monad m => ExplGet m s where
 
 -- | Stores that can be written.
 class Monad m => ExplSet m s where
-  -- | Writes a component to the store.
-  explSet :: s -> Int -> Elem s -> m ()
+  -- | Writes a component to the store. Returns @Just@ a new store if the store
+  -- has changed. If the store is mutable and only undergoes internal mutation,
+  -- then you can safely return @Nothing@
+  explSet :: s -> Int -> Elem s -> m (Maybe s)
```

Comment:
I just realized the `Maybe` is a premature optimization, here and in `explDestroy`. I may just drop it and and return `m s`. That should simplify some other changes.